File: MPPCS/Repository/ShowDatabaseRepository.cs
using CommonDomain;
using Persistence.Repository.Utils;
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Persistence.Repository
{
    public class ShowDatabaseRepository : ShowRepository
    {
        private DbUtils dbUtils;
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        public ShowDatabaseRepository(string url)
        {
            dbUtils = new DbUtils(url);
        }

        public void Add(Show show)
        {
            log.Debug("saving: "+show.GetShowName());
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "INSERT INTO Shows(artist_name,show_name,address,start_time,available_seats,sold_seats) VALUES(@1,@2,@3,@4,@5,@6)";
                command.Parameters.Add(new SQLiteParameter("@1", show.GetArtistName()));
                command.Parameters.Add(new SQLiteParameter("@2", show.GetShowName()));
                command.Parameters.Add(new SQLiteParameter("@3", show.GetAddress()));
                command.Parameters.Add(new SQLiteParameter("@4", show.GetStartTime()));
                command.Parameters.Add(new SQLiteParameter("@5", show.GetAvailableSeats()));
                command.Parameters.Add(new SQLiteParameter("@6",show.GetSoldSeats()));
                int result = command.ExecuteNonQuery();
                log.Debug("SAVED: " + result);
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
        }

        public Show Delete(int id)
        {
            log.Debug("deleting: "+id);
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "DELETE FROM Shows WHERE id = @1";
                command.Parameters.Add(new SQLiteParameter("@1", id));
                Show deleted = Find(id);
                int result = command.ExecuteNonQuery();
                log.Debug("DELETED " + result);
                return deleted;
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
        }

        public Show Find(int id)
        {
            log.Debug("finding: "+id);
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "SELECT * FROM Shows WHERE id = @1";
                command.Parameters.Add(new SQLiteParameter("@1", id));
                SQLiteDataReader dataReader = command.ExecuteReader();
                dataReader.Read();
                Show show = new Show(
                    dataReader.GetString(1),
                    dataReader.GetString(2),
                    dataReader.GetString(3),
                    DateTime.Parse(dataReader.GetString(4)),
                    dataReader.GetInt32(5),
                    dataReader.GetInt32(6));
                show.SetId(dataReader.GetInt32(0));
                dataReader.Close();
                log.Debug("FOUND " + show.GetId());
                return show;
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new Exception(ex.Message);
            }
        }

        public List<Show> GetAll()
        {
            log.Debug("GetAll");
            List<Show> shows = new List<Show>();
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "SELECT * FROM Shows";
                SQLiteDataReader dataReader = command.ExecuteReader();

                while (dataReader.Read())
                {
                    Show show = new Show(
                    dataReader.GetString(1),
                    dataReader.GetString(2),
                    dataReader.GetString(3),
                    DateTime.Parse(dataReader.GetString(4)),
                    dataReader.GetInt32(5),
                    dataReader.GetInt32(6));
                    show.SetId(dataReader.GetInt32(0));
                    shows.Add(show);
                }
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
            log.Debug("GOT ALL "+shows.Count());
            return shows;
        }

        public int Size()
        {
            log.Debug("Size...");
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "SELECT COUNT(*) AS count FROM Shows";
                SQLiteDataReader dataReader = command.ExecuteReader();
                dataReader.Read();
                int size = dataReader.GetInt32(0);
                dataReader.Close();
                log.Debug("SIZE IS "+size);
                return size;
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
        }

        public Show Update(Show show)
        {
            log.Debug("updating: " +show.GetId());
            SQLiteConnection con = dbUtils.GetConnection();
            try
            {
                SQLiteCommand command = con.CreateCommand();
                command.CommandText = "UPDATE Shows SET artist_name=@1, show_name=@2,address=@3,start_time=@4,available_seats=@5,sold_seats=@6 WHERE id=@7";
                command.Parameters.Add(new SQLiteParameter("@1", show.GetArtistName()));
                command.Parameters.Add(new SQLiteParameter("@2", show.GetShowName()));
                command.Parameters.Add(new SQLiteParameter("@3", show.GetAddress()));
                command.Parameters.Add(new SQLiteParameter("@4", show.GetStartTime()));
                command.Parameters.Add(new SQLiteParameter("@5", show.GetAvailableSeats()));
                command.Parameters.Add(new SQLiteParameter("@6", show.GetSoldSeats()));
                command.Parameters.Add(new SQLiteParameter("@7", show.GetId()));
                Show updated = Find(show.GetId());
                int result = command.ExecuteNonQuery();
                log.Debug("UPDATED: " + result);
                return updated;
            }
            catch (Exception ex)
            {
                log.Error(ex);
                throw new NotImplementedException();
            }
        }
    }
}

File: PersistenceAndServices/src/main/java/repository/ShowDatabaseRepository.java
package repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import domain.Show;
import repository.utils.JdbcUtils;

import java.sql.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ShowDatabaseRepository implements ShowRepository{
    private JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public ShowDatabaseRepository (Properties props) {
        logger.info("Initialized ShowDatabaseRepository with properties: {}",props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public Show find(Integer id) throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("SELECT * FROM Shows WHERE id = ?");
            statement.setInt(1,id);
            ResultSet result = statement.executeQuery();
            result.next();
            int found_id = result.getInt("id");
            String artistName = result.getString("artist_name");
            String showName   = result.getString("show_name");
            String address    = result.getString("address");
            LocalDateTime date = LocalDateTime.parse(result.getString("start_time"));
            int availableSeats = result.getInt("available_seats");
            int soldSeats = result.getInt("sold_seats");

            Show show = new Show(artistName,showName,address,date,availableSeats,soldSeats);
            show.setID(found_id);
            result.close();
            logger.traceExit(show.getID()+show.getSoldSeats());
            return show;
        }
        catch (SQLException ex) {
            logger.error(ex);
            throw new Exception("No such ID exists!");
        }
    }

    @Override
    public List<Show> getAll() throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Show> shows = new ArrayList<>();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("SELECT * FROM Shows");
            ResultSet result = statement.executeQuery();
            while (result.next()) {
                int found_id = result.getInt("id");
                String artistName = result.getString("artist_name");
                String showName   = result.getString("show_name");
                String address    = result.getString("address");
                LocalDateTime date = LocalDateTime.parse(result.getString("start_time"));
                int availableSeats = result.getInt("available_seats");
                int soldSeats = result.getInt("sold_seats");

                Show show = new Show(artistName,showName,address,date,availableSeats,soldSeats);
                show.setID(found_id);
                shows.add(show);
            }
            result.close();
        }
        catch (SQLException ex) {
            logger.error(ex);
            throw new Exception("Error getting all sales!");
        }
        logger.traceExit(shows.size());
        return shows;
    }

    @Override
    public void add(Show show) throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("INSERT INTO Shows(artist_name,show_name,address,start_time,available_seats,sold_seats) VALUES(?,?,?,?,?,?)");
            statement.setString(1,show.getArtistName());
            statement.setString(2,show.getShowName());
            statement.setString(3,show.getAddress());
            statement.setString(4, show.getStartTime().toString());
            statement.setInt(5,show.getAvailableSeats());
            statement.setInt(6,show.getSoldSeats());
            int result = statement.executeUpdate();
            logger.trace("Saved {} instance: {}",result,show.getShowName());
        }
        catch (SQLException ex) {
            logger.error(ex);
            throw new Exception("Adding show failed!");
        }
    }

    @Override
    public Show delete(Integer id) throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("DELETE FROM Shows WHERE id = ?");
            statement.setInt(1,id);
            Show deleted = find(id);
            int result = statement.executeUpdate();
            logger.trace("Deleted {} instance",result);
            logger.traceExit(deleted);
            return deleted;
        }
        catch (Exception ex) {
            logger.error(ex);
            throw new Exception("No such ID exists!");
        }
    }

    @Override
    public Show update(Show show) throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("UPDATE Shows SET artist_name=?, show_name=?,address=?,start_time=?,available_seats=?,sold_seats=? WHERE id=?");
            statement.setString(1,show.getArtistName());
            statement.setString(2,show.getShowName());
            statement.setString(3,show.getAddress());
            statement.setString(4, show.getStartTime().toString());
            statement.setInt(5,show.getAvailableSeats());
            statement.setInt(6,show.getSoldSeats());
            statement.setInt(7,show.getID());
            Show updated = find(show.getID());
            int result = statement.executeUpdate();
            logger.trace("Updated {} instances: {} seats sold",result,show.getSoldSeats());
            logger.traceExit(updated);
            return updated;
        }
        catch (Exception ex) {
            logger.error(ex);
            throw new Exception("Updating show failed!");
        }
    }

    @Override
    public Integer size() throws Exception {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try {
            PreparedStatement statement = con.prepareStatement
                    ("SELECT COUNT(*) AS count FROM Shows");
            ResultSet result = statement.executeQuery();
            result.next();
            int size = result.getInt("count");
            result.close();
            logger.traceExit(size);
            return size;
        }
        catch (Exception ex) {
            logger.error(ex);
            throw new Exception("Error in database!");
        }
    }
}
